File: src/main/java/com/dauren/storage/SqliteStorage.java
package com.dauren.storage;

import com.dauren.model.TaskType;

import java.sql.*;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

public class SqliteStorage {
    private final String jdbcUrl;

    public SqliteStorage(String jdbcUrl) {
        this.jdbcUrl = jdbcUrl;
    }

    private Connection connection() throws SQLException {
        return DriverManager.getConnection(jdbcUrl);
    }

    public void ensureUser(String userId, String userName) {
        String now = Instant.now().toString();
        try (Connection c = connection()) {
            try (PreparedStatement ps = c.prepareStatement("INSERT OR IGNORE INTO users(id, whatsapp_name, created_at) VALUES(?,?,?)")) {
                ps.setString(1, userId);
                ps.setString(2, userName);
                ps.setString(3, now);
                ps.executeUpdate();
            }
            try (PreparedStatement ps = c.prepareStatement("UPDATE users SET whatsapp_name=? WHERE id=? AND ? IS NOT NULL")) {
                ps.setString(1, userName);
                ps.setString(2, userId);
                ps.setString(3, userName);
                ps.executeUpdate();
            }
            try (PreparedStatement ps = c.prepareStatement("INSERT OR IGNORE INTO profiles(user_id, surname, role) VALUES(?, NULL, NULL)")) {
                ps.setString(1, userId);
                ps.executeUpdate();
            }
            try (PreparedStatement ps = c.prepareStatement("INSERT OR IGNORE INTO access_levels(user_id, level, updated_at) VALUES(?,1,?)")) {
                ps.setString(1, userId);
                ps.setString(2, now);
                ps.executeUpdate();
            }
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    public void rememberGroupMember(String chatId, String userId) {
        try (Connection c = connection(); PreparedStatement ps = c.prepareStatement("INSERT OR IGNORE INTO group_members(chat_id,user_id) VALUES(?,?)")) {
            ps.setString(1, chatId);
            ps.setString(2, userId);
            ps.executeUpdate();
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    public int getAccessLevel(String userId) {
        try (Connection c = connection(); PreparedStatement ps = c.prepareStatement("SELECT level FROM access_levels WHERE user_id=?")) {
            ps.setString(1, userId);
            try (ResultSet rs = ps.executeQuery()) {
                return rs.next() ? rs.getInt(1) : 1;
            }
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    public void setAccessLevel(String userId, int level) {
        String now = Instant.now().toString();
        try (Connection c = connection(); PreparedStatement ps = c.prepareStatement("INSERT INTO access_levels(user_id, level, updated_at) VALUES(?,?,?) ON CONFLICT(user_id) DO UPDATE SET level=excluded.level, updated_at=excluded.updated_at")) {
            ps.setString(1, userId);
            ps.setInt(2, level);
            ps.setString(3, now);
            ps.executeUpdate();
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    public Optional<UserProfileRow> getProfile(String userId) {
        String sql = "SELECT u.id, u.whatsapp_name, p.surname, p.role, COALESCE(a.level,1) AS access_level FROM users u LEFT JOIN profiles p ON p.user_id=u.id LEFT JOIN access_levels a ON a.user_id=u.id WHERE u.id=?";
        try (Connection c = connection(); PreparedStatement ps = c.prepareStatement(sql)) {
            ps.setString(1, userId);
            try (ResultSet rs = ps.executeQuery()) {
                if (!rs.next()) return Optional.empty();
                return Optional.of(new UserProfileRow(rs.getString("id"), rs.getString("whatsapp_name"), rs.getString("surname"), rs.getString("role"), rs.getInt("access_level")));
            }
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    public List<UserProfileRow> listGroupProfiles(String chatId) {
        String sql = "SELECT u.id, u.whatsapp_name, p.surname, p.role, COALESCE(a.level,1) AS access_level FROM group_members gm JOIN users u ON u.id=gm.user_id LEFT JOIN profiles p ON p.user_id=u.id LEFT JOIN access_levels a ON a.user_id=u.id WHERE gm.chat_id=? ORDER BY COALESCE(p.surname,u.whatsapp_name,u.id)";
        List<UserProfileRow> rows = new ArrayList<>();
        try (Connection c = connection(); PreparedStatement ps = c.prepareStatement(sql)) {
            ps.setString(1, chatId);
            try (ResultSet rs = ps.executeQuery()) {
                while (rs.next()) {
                    rows.add(new UserProfileRow(rs.getString("id"), rs.getString("whatsapp_name"), rs.getString("surname"), rs.getString("role"), rs.getInt("access_level")));
                }
            }
            return rows;
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    public void setSurname(String userId, String surname) { updateProfile(userId, surname, null); }
    public void setRole(String userId, String role) { updateProfile(userId, null, role); }

    public void setProfile(String userId, String surname, String role) { updateProfile(userId, surname, role); }

    private void updateProfile(String userId, String surname, String role) {
        ensureUser(userId, null);
        try (Connection c = connection()) {
            if (surname != null) {
                try (PreparedStatement ps = c.prepareStatement("UPDATE profiles SET surname=? WHERE user_id=?")) {
                    ps.setString(1, surname);
                    ps.setString(2, userId);
                    ps.executeUpdate();
                }
            }
            if (role != null) {
                try (PreparedStatement ps = c.prepareStatement("UPDATE profiles SET role=? WHERE user_id=?")) {
                    ps.setString(1, role);
                    ps.setString(2, userId);
                    ps.executeUpdate();
                }
            }
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    public long createTask(String chatId, String assigneeId, String createdBy, String text, TaskType type, String dueDate, String weekday, String remindTime) {
        String now = Instant.now().toString();
        String sql = "INSERT INTO tasks(chat_id, assignee_id, created_by, text, type, due_date, weekday, remind_time, created_at) VALUES(?,?,?,?,?,?,?,?,?)";
        try (Connection c = connection(); PreparedStatement ps = c.prepareStatement(sql, Statement.RETURN_GENERATED_KEYS)) {
            ps.setString(1, chatId); ps.setString(2, assigneeId); ps.setString(3, createdBy); ps.setString(4, text);
            ps.setString(5, type.name().toLowerCase()); ps.setString(6, dueDate); ps.setString(7, weekday); ps.setString(8, remindTime); ps.setString(9, now);
            ps.executeUpdate();
            try (ResultSet rs = ps.getGeneratedKeys()) {
                return rs.next() ? rs.getLong(1) : -1L;
            }
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    public List<TaskRow> listActiveTasks(String chatId) {
        String sql = "SELECT * FROM tasks WHERE chat_id=? AND deleted_at IS NULL AND (done_at IS NULL OR type='weekly') ORDER BY id";
        List<TaskRow> rows = new ArrayList<>();
        try (Connection c = connection(); PreparedStatement ps = c.prepareStatement(sql)) {
            ps.setString(1, chatId);
            try (ResultSet rs = ps.executeQuery()) {
                while (rs.next()) rows.add(TaskRow.from(rs));
            }
            return rows;
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    public Optional<TaskRow> getTask(long taskId) {
        try (Connection c = connection(); PreparedStatement ps = c.prepareStatement("SELECT * FROM tasks WHERE id=?")) {
            ps.setLong(1, taskId);
            try (ResultSet rs = ps.executeQuery()) {
                return rs.next() ? Optional.of(TaskRow.from(rs)) : Optional.empty();
            }
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    public void markTaskDone(long taskId, String cycleKey, String actorId) {
        String now = Instant.now().toString();
        try (Connection c = connection()) {
            try (PreparedStatement ps = c.prepareStatement("UPDATE tasks SET done_at=? WHERE id=? AND type!='weekly'")) {
                ps.setString(1, now); ps.setLong(2, taskId); ps.executeUpdate();
            }
            try (PreparedStatement ps = c.prepareStatement("INSERT INTO task_events(task_id,event_type,actor_id,payload,event_at) VALUES(?,?,?,?,?)")) {
                ps.setLong(1, taskId);
                ps.setString(2, "done");
                ps.setString(3, actorId);
                ps.setString(4, "{\"cycle\":\"" + cycleKey + "\"}");
                ps.setString(5, now);
                ps.executeUpdate();
            }
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    public void deleteTask(long taskId) {
        try (Connection c = connection(); PreparedStatement ps = c.prepareStatement("UPDATE tasks SET deleted_at=? WHERE id=?")) {
            ps.setString(1, Instant.now().toString());
            ps.setLong(2, taskId);
            ps.executeUpdate();
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    public List<TaskRow> dueTasks(String date, String time, String weekday) {
        String sql = "SELECT * FROM tasks WHERE deleted_at IS NULL AND ((type='dated' AND done_at IS NULL AND due_date=? AND remind_time<=?) OR (type='weekly' AND weekday=? AND remind_time<=?))";
        List<TaskRow> rows = new ArrayList<>();
        try (Connection c = connection(); PreparedStatement ps = c.prepareStatement(sql)) {
            ps.setString(1, date); ps.setString(2, time); ps.setString(3, weekday); ps.setString(4, time);
            try (ResultSet rs = ps.executeQuery()) {
                while (rs.next()) rows.add(TaskRow.from(rs));
            }
            return rows;
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    public boolean hasEvent(long taskId, String eventType, String cycleKey) {
        String sql = "SELECT id FROM task_events WHERE task_id=? AND event_type=? AND payload LIKE ? LIMIT 1";
        try (Connection c = connection(); PreparedStatement ps = c.prepareStatement(sql)) {
            ps.setLong(1, taskId); ps.setString(2, eventType); ps.setString(3, "%\"cycle\":\"" + cycleKey + "\"%");
            try (ResultSet rs = ps.executeQuery()) { return rs.next(); }
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    public void addTaskEvent(long taskId, String eventType, String payload) {
        try (Connection c = connection(); PreparedStatement ps = c.prepareStatement("INSERT INTO task_events(task_id,event_type,payload,event_at) VALUES(?,?,?,?)")) {
            ps.setLong(1, taskId);
            ps.setString(2, eventType);
            ps.setString(3, payload);
            ps.setString(4, Instant.now().toString());
            ps.executeUpdate();
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    public void saveContext(String scopeType, String scopeId, String contextJson, boolean greeted) {
        String sql = "INSERT INTO conversation_contexts(scope_type,scope_id,context_json,greeted,updated_at) VALUES(?,?,?,?,?) ON CONFLICT(scope_type,scope_id) DO UPDATE SET context_json=excluded.context_json,greeted=excluded.greeted,updated_at=excluded.updated_at";
        try (Connection c = connection(); PreparedStatement ps = c.prepareStatement(sql)) {
            ps.setString(1, scopeType); ps.setString(2, scopeId); ps.setString(3, contextJson); ps.setInt(4, greeted ? 1 : 0); ps.setString(5, Instant.now().toString());
            ps.executeUpdate();
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    public Optional<ContextRow> getContext(String scopeType, String scopeId) {
        try (Connection c = connection(); PreparedStatement ps = c.prepareStatement("SELECT * FROM conversation_contexts WHERE scope_type=? AND scope_id=?")) {
            ps.setString(1, scopeType); ps.setString(2, scopeId);
            try (ResultSet rs = ps.executeQuery()) {
                return rs.next() ? Optional.of(new ContextRow(rs.getString("scope_type"), rs.getString("scope_id"), rs.getString("context_json"), rs.getInt("greeted") == 1)) : Optional.empty();
            }
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    public void audit(String actorId, String action, String objectType, String objectId, String detailsJson) {
        try (Connection c = connection(); PreparedStatement ps = c.prepareStatement("INSERT INTO audit_logs(actor_id,action,object_type,object_id,details,at) VALUES(?,?,?,?,?,?)")) {
            ps.setString(1, actorId); ps.setString(2, action); ps.setString(3, objectType); ps.setString(4, objectId); ps.setString(5, detailsJson); ps.setString(6, Instant.now().toString());
            ps.executeUpdate();
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    public Optional<RateLimitRow> getRate(String key) {
        try (Connection c = connection(); PreparedStatement ps = c.prepareStatement("SELECT * FROM rate_limits WHERE key=?")) {
            ps.setString(1, key);
            try (ResultSet rs = ps.executeQuery()) {
                return rs.next() ? Optional.of(new RateLimitRow(rs.getString("key"), rs.getLong("window_start"), rs.getInt("count"), rs.getLong("cooldown_until"))) : Optional.empty();
            }
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    public void upsertRate(String key, long windowStart, int count, long cooldownUntil) {
        String sql = "INSERT INTO rate_limits(key,window_start,count,cooldown_until) VALUES(?,?,?,?) ON CONFLICT(key) DO UPDATE SET window_start=excluded.window_start,count=excluded.count,cooldown_until=excluded.cooldown_until";
        try (Connection c = connection(); PreparedStatement ps = c.prepareStatement(sql)) {
            ps.setString(1, key); ps.setLong(2, windowStart); ps.setInt(3, count); ps.setLong(4, cooldownUntil); ps.executeUpdate();
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    public record UserProfileRow(String id, String whatsappName, String surname, String role, int accessLevel) {}
    public record ContextRow(String scopeType, String scopeId, String contextJson, boolean greeted) {}
    public record RateLimitRow(String key, long windowStart, int count, long cooldownUntil) {}

    public record TaskRow(long id, String chatId, String assigneeId, String createdBy, String text, String type,
                          String dueDate, String weekday, String remindTime, String doneAt, String deletedAt) {
        static TaskRow from(ResultSet rs) throws SQLException {
            return new TaskRow(
                    rs.getLong("id"), rs.getString("chat_id"), rs.getString("assignee_id"), rs.getString("created_by"), rs.getString("text"), rs.getString("type"),
                    rs.getString("due_date"), rs.getString("weekday"), rs.getString("remind_time"), rs.getString("done_at"), rs.getString("deleted_at")
            );
        }

        public TaskType taskType() {
            return TaskType.valueOf(type.toUpperCase());
        }
    }
}

File: legacy-node/src-js/repositories/sqliteStorage.js
import { execFileSync } from 'node:child_process';
import { mkdirSync } from 'node:fs';
import { dirname } from 'node:path';

function esc(value) {
  if (value === null || value === undefined) return 'NULL';
  if (typeof value === 'number') return String(value);
  if (typeof value === 'boolean') return value ? '1' : '0';
  return `'${String(value).replace(/'/g, "''")}'`;
}

export class SqliteStorage {
  constructor(dbPath = '.data/dauren.sqlite') {
    this.dbPath = dbPath;
    mkdirSync(dirname(dbPath), { recursive: true });
    this.initSchema();
  }

  run(sql) {
    execFileSync('sqlite3', [this.dbPath, sql], { encoding: 'utf-8' });
  }

  all(sql) {
    const out = execFileSync('sqlite3', ['-json', this.dbPath, sql], { encoding: 'utf-8' });
    return out.trim() ? JSON.parse(out) : [];
  }

  get(sql) {
    return this.all(sql)[0] ?? null;
  }

  initSchema() {
    this.run(`
      PRAGMA journal_mode=WAL;
      CREATE TABLE IF NOT EXISTS users (
        id TEXT PRIMARY KEY,
        whatsapp_name TEXT,
        created_at TEXT NOT NULL
      );
      CREATE TABLE IF NOT EXISTS profiles (
        user_id TEXT PRIMARY KEY,
        surname TEXT,
        role TEXT,
        FOREIGN KEY(user_id) REFERENCES users(id)
      );
      CREATE TABLE IF NOT EXISTS access_levels (
        user_id TEXT PRIMARY KEY,
        level INTEGER NOT NULL DEFAULT 1,
        updated_at TEXT NOT NULL,
        FOREIGN KEY(user_id) REFERENCES users(id)
      );
      CREATE TABLE IF NOT EXISTS group_members (
        chat_id TEXT NOT NULL,
        user_id TEXT NOT NULL,
        PRIMARY KEY(chat_id, user_id)
      );
      CREATE TABLE IF NOT EXISTS tasks (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        chat_id TEXT NOT NULL,
        assignee_id TEXT NOT NULL,
        created_by TEXT NOT NULL,
        text TEXT NOT NULL,
        type TEXT NOT NULL,
        due_date TEXT,
        weekday TEXT,
        remind_time TEXT,
        done_at TEXT,
        deleted_at TEXT,
        created_at TEXT NOT NULL
      );
      CREATE TABLE IF NOT EXISTS task_events (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        task_id INTEGER NOT NULL,
        event_type TEXT NOT NULL,
        actor_id TEXT,
        payload TEXT,
        event_at TEXT NOT NULL,
        FOREIGN KEY(task_id) REFERENCES tasks(id)
      );
      CREATE TABLE IF NOT EXISTS audit_logs (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        actor_id TEXT,
        action TEXT NOT NULL,
        object_type TEXT,
        object_id TEXT,
        details TEXT,
        at TEXT NOT NULL
      );
      CREATE TABLE IF NOT EXISTS conversation_contexts (
        scope_type TEXT NOT NULL,
        scope_id TEXT NOT NULL,
        context_json TEXT,
        greeted INTEGER NOT NULL DEFAULT 0,
        updated_at TEXT NOT NULL,
        PRIMARY KEY(scope_type, scope_id)
      );
      CREATE TABLE IF NOT EXISTS rate_limits (
        key TEXT PRIMARY KEY,
        window_start INTEGER,
        count INTEGER,
        cooldown_until INTEGER
      );
    `);
  }

  ensureUser(userId, whatsappName = null) {
    const now = new Date().toISOString();
    this.run(`INSERT OR IGNORE INTO users(id, whatsapp_name, created_at) VALUES(${esc(userId)}, ${esc(whatsappName)}, ${esc(now)});`);
    if (whatsappName) {
      this.run(`UPDATE users SET whatsapp_name=${esc(whatsappName)} WHERE id=${esc(userId)};`);
    }
    this.run(`INSERT OR IGNORE INTO profiles(user_id) VALUES(${esc(userId)});`);
    this.run(`INSERT OR IGNORE INTO access_levels(user_id, level, updated_at) VALUES(${esc(userId)}, 1, ${esc(now)});`);
    return this.getUser(userId);
  }

  rememberGroupMember(chatId, userId) {
    this.run(`INSERT OR IGNORE INTO group_members(chat_id, user_id) VALUES(${esc(chatId)}, ${esc(userId)});`);
  }

  listGroupMembers(chatId) {
    return this.all(`
      SELECT u.id, u.whatsapp_name, p.surname, p.role, a.level as access_level
      FROM group_members gm
      JOIN users u ON u.id = gm.user_id
      LEFT JOIN profiles p ON p.user_id = u.id
      LEFT JOIN access_levels a ON a.user_id = u.id
      WHERE gm.chat_id = ${esc(chatId)}
      ORDER BY COALESCE(p.surname, u.whatsapp_name, u.id) COLLATE NOCASE;
    `);
  }

  getUser(userId) {
    return this.get(`
      SELECT u.id, u.whatsapp_name, p.surname, p.role, COALESCE(a.level,1) as access_level
      FROM users u
      LEFT JOIN profiles p ON p.user_id = u.id
      LEFT JOIN access_levels a ON a.user_id = u.id
      WHERE u.id = ${esc(userId)};
    `);
  }

  setProfile(userId, { surname, role }) {
    if (surname !== undefined) this.run(`UPDATE profiles SET surname=${esc(surname)} WHERE user_id=${esc(userId)};`);
    if (role !== undefined) this.run(`UPDATE profiles SET role=${esc(role)} WHERE user_id=${esc(userId)};`);
    return this.getUser(userId);
  }

  getAccessLevel(userId) {
    const row = this.get(`SELECT level FROM access_levels WHERE user_id=${esc(userId)};`);
    return row ? Number(row.level) : 1;
  }

  setAccessLevel(userId, level) {
    const now = new Date().toISOString();
    this.run(`INSERT INTO access_levels(user_id, level, updated_at) VALUES(${esc(userId)}, ${esc(level)}, ${esc(now)})
      ON CONFLICT(user_id) DO UPDATE SET level=excluded.level, updated_at=excluded.updated_at;`);
  }

  createTask(task) {
    const now = new Date().toISOString();
    this.run(`
      INSERT INTO tasks(chat_id, assignee_id, created_by, text, type, due_date, weekday, remind_time, created_at)
      VALUES(${esc(task.chatId)}, ${esc(task.assigneeId)}, ${esc(task.createdBy)}, ${esc(task.text)}, ${esc(task.type)},
             ${esc(task.dueDate)}, ${esc(task.weekday)}, ${esc(task.remindTime)}, ${esc(now)});
    `);
    return this.get('SELECT * FROM tasks ORDER BY id DESC LIMIT 1;');
  }

  listActiveTasks(chatId) {
    return this.all(`SELECT * FROM tasks WHERE chat_id=${esc(chatId)} AND deleted_at IS NULL AND done_at IS NULL ORDER BY id ASC;`);
  }

  getTask(taskId) {
    return this.get(`SELECT * FROM tasks WHERE id=${esc(taskId)};`);
  }

  markTaskDone(taskId, doneAt = new Date().toISOString()) {
    this.run(`UPDATE tasks SET done_at=${esc(doneAt)} WHERE id=${esc(taskId)};`);
  }

  deleteTask(taskId, deletedAt = new Date().toISOString()) {
    this.run(`UPDATE tasks SET deleted_at=${esc(deletedAt)} WHERE id=${esc(taskId)};`);
  }

  logTaskEvent(taskId, eventType, actorId = null, payload = null) {
    const now = new Date().toISOString();
    this.run(`INSERT INTO task_events(task_id, event_type, actor_id, payload, event_at) VALUES(${esc(taskId)}, ${esc(eventType)}, ${esc(actorId)}, ${esc(payload ? JSON.stringify(payload) : null)}, ${esc(now)});`);
  }

  hasTaskEvent(taskId, eventType, periodKey) {
    const row = this.get(`SELECT id FROM task_events WHERE task_id=${esc(taskId)} AND event_type=${esc(eventType)} AND payload LIKE ${esc(`%"periodKey":"${periodKey}"%`)} LIMIT 1;`);
    return Boolean(row);
  }

  getDueTasks(dateStr, timeStr, weekday) {
    return this.all(`
      SELECT * FROM tasks
      WHERE deleted_at IS NULL
        AND done_at IS NULL
        AND (
          (type='dated' AND due_date=${esc(dateStr)} AND remind_time <= ${esc(timeStr)})
          OR
          (type='weekly' AND weekday=${esc(weekday)} AND remind_time <= ${esc(timeStr)})
        );
    `);
  }

  writeAudit({ actorId = null, action, objectType = null, objectId = null, details = null }) {
    const at = new Date().toISOString();
    this.run(`INSERT INTO audit_logs(actor_id, action, object_type, object_id, details, at)
              VALUES(${esc(actorId)}, ${esc(action)}, ${esc(objectType)}, ${esc(objectId)}, ${esc(details ? JSON.stringify(details) : null)}, ${esc(at)});`);
  }

  getContext(scopeType, scopeId) {
    return this.get(`SELECT * FROM conversation_contexts WHERE scope_type=${esc(scopeType)} AND scope_id=${esc(scopeId)};`);
  }

  saveContext(scopeType, scopeId, contextJson, greeted) {
    const now = new Date().toISOString();
    this.run(`INSERT INTO conversation_contexts(scope_type, scope_id, context_json, greeted, updated_at)
      VALUES(${esc(scopeType)}, ${esc(scopeId)}, ${esc(JSON.stringify(contextJson || []))}, ${esc(greeted ? 1 : 0)}, ${esc(now)})
      ON CONFLICT(scope_type, scope_id)
      DO UPDATE SET context_json=excluded.context_json, greeted=excluded.greeted, updated_at=excluded.updated_at;`);
  }

  clearContext(scopeType, scopeId) {
    const now = new Date().toISOString();
    this.run(`INSERT INTO conversation_contexts(scope_type, scope_id, context_json, greeted, updated_at)
      VALUES(${esc(scopeType)}, ${esc(scopeId)}, ${esc('[]')}, 1, ${esc(now)})
      ON CONFLICT(scope_type, scope_id)
      DO UPDATE SET context_json='[]', updated_at=excluded.updated_at;`);
  }

  getRateLimit(key) {
    return this.get(`SELECT * FROM rate_limits WHERE key=${esc(key)};`);
  }

  saveRateLimit(key, windowStart, count, cooldownUntil = 0) {
    this.run(`INSERT INTO rate_limits(key, window_start, count, cooldown_until)
      VALUES(${esc(key)}, ${esc(windowStart)}, ${esc(count)}, ${esc(cooldownUntil)})
      ON CONFLICT(key) DO UPDATE SET window_start=excluded.window_start, count=excluded.count, cooldown_until=excluded.cooldown_until;`);
  }
}
